### ISSUE TITLE
Creating continuation inside of ring

### ISSUE THREAD

**Issue by ishanpm** [jmoenig/Snap#2319]
There seems to be some undefined behavior when creating a continuation inside of a ring and then calling it outside of the ring. I have no idea what the intended behavior should be, but it currently acts differently based on things that shouldn't matter. This script, for example, acts in different ways if some of the "ask" blocks are removed:

![snap cc glitch](https://user-images.githubusercontent.com/10391397/52179276-b1eda100-279d-11e9-882a-2fe3b780c7b4.png)

- If the script is run as shown, the sprite will ask a, b, c, b, d.
- If the "ask d" block is removed, then the sprite will ask a, b, c, b, b, b, b, b... When running in single stepping mode, it looks like evaluation continues from the "run w/ continuation" block when exiting the ring, causing an infinite loop.
- If the "ask b" block is removed, then the sprite will ask a, c, c, c, c... This happens whether or not the "ask d" block is present.
- Also, just for completeness: If the code inside the run block is inlined to the surrounding code, the sprite will ask a, b, c, b, c, b, c...

This isn't a terribly high-priority issue, but it complicates some custom control structures (which I know aren't completely supported anyway).

Possibly related to this issue, since they're both about continuations returning to weird places: #2091

**Comment by rdococ** [comment]
The correct sequence of events here is a, b, c, b, c, b, c..., because running a continuation usually means that you don't expect it to report back like a regular procedure would. In no scenario should the "ask d" block ever be evaluated, because the continuation will always be invoked before it. To be honest, I wish Snap! made more of a distinction between continuations and procedures, but that's another issue.

I'm beginning to wonder how exactly continuations were implemented.

**Comment by brianharvey** [comment]
Yes, well, you're probably the only person besides me who's ever played with it, so we haven't had a lot of bug reports (except indirectly, when people complain about CATCH and THROW being broken).

We know about the problem of a run/cc in tail position.  (Look at the definition of command CATCH in the tools library. :-) )  It's just never been the most urgent thing to fix.

If you have a spaghetti stack to begin with, implementing continuations in a text language should just mean saving the stack pointer and the environment pointer.  For us it's a little more complicated because we try to display the continuation in some pedagogically helpful way.

I asked Jens to implement continuations largely because I wanted to watch Mitch Resnick's head explode when I showed it to him.  It didn't work; his head didn't explode.  Later I asked him why not, and he said "my head would explode if I saw that _in Scratch,_ but you can do anything you want in Snap<i>!</i>."

Then it turned out to be useful in creating CATCH and THROW, which was important because we couldn't find a behavior for a REPORT inside a FOR loop inside a reporter that was The Right Thing in every use case.

Anyway, to sum up this rambling comment, we know continuations are broken and we'll fix them When Things Slow Down™.

**Comment by ishanpm** [comment]
Yeah, I figured it wouldn't really be a high priority thing. I just posted the issue because I ran into it when I was trying to implement coroutines, but I can work around it.
